apps/api/src/notifications/notifications.service.ts

```diff
@@ -0,0 +1,314 @@
+            where: { email: data.email },
+            create: { mailHash: secret, email: data.email },
+            update: { mailHash: secret },
+          })
+
+      // Public subscribe link
+      const link =
+        this.config.get<string>('APP_URL', '') +
+        `/notifications/subscribe?hash=${secret}&email=${data.email}&consent=true`
+
+      // Prepare Email data
+      const recepient = { to: [data.email] }
+      const mail = new ConfirmConsentEmailDto({
+        subscribeLink: link,
+      })
+
+      await this.sendEmail.sendFromTemplate(mail, recepient)
+
+      sent
+        ? // Update new send date
+          await this.prisma.emailSentRegistry.update({
+            where: { id: sent.id },
+            data: { dateSent: new Date() },
+          })
+        : // Create new record for the email sent
+          await this.prisma.emailSentRegistry.create({
+            data: { email: data.email, type: EmailType.confirmConsent, dateSent: new Date() },
+          })
+    } catch (e) {
+      return new BadRequestException('Failed to send email ')
+    }
+
+    return { message: 'Email Sent' }
+  }
+
+  async sendUnregisteredConfirmEmail(data: { email: string; campaignId?: string }) {
+    // Send confirmation email
+    try {
+      const sent = await this.prisma.emailSentRegistry.findFirst({
+        where: { email: data.email, type: EmailType.confirmConsent },
+      })
+      // Check if email was sent already in the past 1 minutes
+      const wasSent = this.emailWasSentRecently(sent, 1)
+
+      if (wasSent) return
+
+      // Secret hash to verify data in the subscribe route
+      const secret = this.generateHash()
+
+      // Save hash to record
+      await this.prisma.unregisteredNotificationConsent.upsert({
+        where: { email: data.email },
+        create: { mailHash: secret, email: data.email },
+        update: { mailHash: secret },
+      })
+
+      // Public subscribe link
+      let link =
+        this.config.get<string>('APP_URL', '') +
+        `/notifications/subscribe?hash=${secret}&email=${data.email}&consent=true`
+      if (data.campaignId) link += `&campaign=${data.campaignId}`
+
+      // Prepare Email data
+      const recepient = { to: [data.email] }
+      const mail = new ConfirmConsentEmailDto({
+        subscribeLink: link,
+      })
+
+      await this.sendEmail.sendFromTemplate(mail, recepient)
+
+      sent
+        ? // Update new send date
+          await this.prisma.emailSentRegistry.update({
+            where: { id: sent.id },
+            data: { dateSent: new Date() },
+          })
+        : // Create new record for the email sent
+          await this.prisma.emailSentRegistry.create({
+            data: { email: data.email, type: EmailType.confirmConsent, dateSent: new Date() },
+          })
+    } catch (e) {
+      return new BadRequestException('Failed to send email ')
+    }
+  }
+
+  async subscribePublic(data: SubscribePublicDto) {
```

Comment:
please add a type for the returned object

Reply:
Done